File: libs/python/edgecommons/messaging/message.py
# (c) 2022 Amazon Web Services, Inc. or its affiliates. All Rights Reserved.
# This AWS Content is provided subject to the terms of the AWS Customer Agreement
# available at http://aws.amazon.com/agreement or other written agreement between
# Customer and Amazon Web Services, Inc.

import base64
import binascii
import json
import logging
from dataclasses import dataclass, field
from uuid import uuid4
from typing import Any, Optional, TYPE_CHECKING

from edgecommons.messaging.identity import MessageIdentity
from edgecommons.utils import Utils


MAX_BINARY_BODY_BYTES = 64 * 1024
BINARY_BODY_KEY = "_edgecommonsBinary"
BINARY_ENCODING = "base64"


def _binary_marker(data: bytes) -> dict:
    if len(data) > MAX_BINARY_BODY_BYTES:
        raise ValueError(f"Binary message body exceeds {MAX_BINARY_BODY_BYTES} bytes")
    return {
        BINARY_BODY_KEY: {
            "encoding": BINARY_ENCODING,
            "length": len(data),
            "data": base64.b64encode(data).decode("ascii"),
        }
    }


def _encode_body(body: Any) -> Any:
    """Encode a top-level binary body as the first-class bounded binary marker."""
    if isinstance(body, (bytes, bytearray)):
        return _binary_marker(bytes(body))
    return body


def _binary_descriptor(value: Any) -> Optional[dict]:
    if isinstance(value, dict):
        if BINARY_BODY_KEY not in value:
            return None
        descriptor = value[BINARY_BODY_KEY]
        if not isinstance(descriptor, dict):
            raise ValueError("Binary message body marker must be an object")
        return descriptor
    return None


def _decode_binary_descriptor(descriptor: dict) -> bytes:
    if descriptor.get("encoding") != BINARY_ENCODING:
        raise ValueError("Binary message body encoding must be base64")
    declared_length = descriptor.get("length")
    if not isinstance(declared_length, int) or declared_length < 0:
        raise ValueError("Binary message body length must be a non-negative integer")
    if declared_length > MAX_BINARY_BODY_BYTES:
        raise ValueError(f"Binary message body exceeds {MAX_BINARY_BODY_BYTES} bytes")
    encoded = descriptor.get("data")
    if not isinstance(encoded, str):
        raise ValueError("Binary message body data is required")
    try:
        decoded = base64.b64decode(encoded, validate=True)
    except (binascii.Error, ValueError) as exc:
        raise ValueError("Binary message body data is not valid base64") from exc
    if len(decoded) != declared_length:
        raise ValueError("Binary message body length does not match decoded data")
    return decoded


if TYPE_CHECKING:
    from edgecommons.config.manager.config_manager import ConfigManager


logger = logging.getLogger("Message")


@dataclass
class MessageHeader:
    REPLY_MESSAGE_TOPIC_PREFIX = "edgecommons/reply-"  # class constant, not a field

    name: str
    version: str
    correlation_id: Optional[str] = None
    timestamp: Optional[str] = None
    uuid: Optional[str] = None
    reply_to: Optional[str] = None

    def __post_init__(self):
        # Fill computed defaults (matches the previous hand-written constructor).
        if self.timestamp is None:
            self.timestamp = Utils.get_utc_z()
        if self.correlation_id is None:
            self.correlation_id = str(uuid4())
        if self.uuid is None:
            self.uuid = str(uuid4())

    @staticmethod
    def from_dict(src: dict):
        name = src.get("name")
        version = src.get("version")
        timestamp = src.get("timestamp")
        uuid = src.get("uuid")
        correlation_id = src.get("correlation_id")
        reply_to = src.get("reply_to")
        return MessageHeader(name, version, correlation_id, timestamp, uuid, reply_to)

    def to_dict(self) -> dict:
        header = {
            "name": self.name,
            "version": self.version,
            "timestamp": self.timestamp,
            "uuid": self.uuid,
            "correlation_id": self.correlation_id,
        }
        if self.reply_to is not None:
            header["reply_to"] = self.reply_to
        return header

    def make_request(self, reply_to: str = None) -> str:
        if reply_to is None:
            reply_to = self.REPLY_MESSAGE_TOPIC_PREFIX + str(uuid4())
        self.reply_to = reply_to
        logger.debug(f"Setting replyTo field as {self.reply_to}")
        return self.reply_to

    def get_reply_to(self) -> str:
        return self.reply_to

    def set_correlation_id(self, correlation_id: str):
        self.correlation_id = correlation_id


@dataclass
class MessageTags:
    """Free-form message metadata tags.

    The legacy ``thing`` special-casing is removed (UNS-CANONICAL-DESIGN §1.1 — hard
    cut): the publisher's device now travels in the top-level ``identity`` element. A
    stray inbound ``thing`` key just lands in the generic tag map — no legacy shim.
    """

    tags: dict = field(default_factory=dict)

    def __post_init__(self):
        if self.tags is None:
            self.tags = {}

    @staticmethod
    def from_config(config_service: 'ConfigManager'):
        tag_config = config_service.get_tag_config()
        if tag_config is not None:
            return MessageTags(tag_config.to_dict())
        else:
            return MessageTags({})

    @staticmethod
    def from_dict(src: dict):
        return MessageTags(dict(src))

    def inject_tag(self, key: str, value: str):
        self.tags[key] = value

    def to_dict(self) -> dict:
        return dict(self.tags)


@dataclass
class Message:
    header: Optional[MessageHeader] = None
    identity: Optional[MessageIdentity] = None
    tags: Optional[MessageTags] = None
    body: Any = None
    raw: Any = None

    def to_dict(self) -> dict:
        if self.raw is None:
            msg = {}
            if self.header is not None:
                msg["header"] = self.header.to_dict()
            # Canonical envelope member order: header, identity, tags, body.
            if self.identity is not None:
                msg["identity"] = self.identity.to_dict()
            if self.tags is not None:
                msg["tags"] = self.tags.to_dict()
            msg["body"] = _encode_body(self.body)
            return msg
        else:
            return {"raw": _encode_body(self.raw)}

    def __str__(self) -> str:
        return json.dumps(self.to_dict())

    def dumps(self, indent=None) -> str:
        msg = {}
        if self.header is not None:
            msg["header"] = self.header.to_dict()
        if self.identity is not None:
            msg["identity"] = self.identity.to_dict()
        if self.tags is not None:
            msg["tags"] = self.tags.to_dict()
        msg["body"] = _encode_body(self.body)
        return json.dumps(msg, indent=indent)

    def get_correlation_id(self) -> str:
        if self.header is None:
            return None
        return self.header.correlation_id

    def get_header(self) -> MessageHeader:
        return self.header

    def get_identity(self) -> Optional[MessageIdentity]:
        """The UNS identity element of this message (``hier``/``path``/``component``/
        ``instance``), or ``None`` when the message carries none (raw messages,
        messages built without a config-bound builder, or a malformed inbound
        identity)."""
        return self.identity

    def get_tags(self) -> MessageTags:
        return self.tags

    def get_source(self):
        """Backward compatibility alias for get_tags()"""
        return self.get_tags()

    def inject_tag(self, key: str, value: str):
        if self.tags is None:
            self.tags = MessageTags(None)
        self.tags.inject_tag(key, value)

    def get_body(self):
        return self.body

    def get_payload(self):
        """Backward compatibility alias for get_body()"""
        return self.get_body()

    def is_binary_body(self) -> bool:
        return isinstance(self.body, (bytes, bytearray)) or (
            isinstance(self.body, dict) and BINARY_BODY_KEY in self.body
        )

    def get_binary_body(self) -> Optional[bytes]:
        if isinstance(self.body, (bytes, bytearray)):
            data = bytes(self.body)
            if len(data) > MAX_BINARY_BODY_BYTES:
                raise ValueError(f"Binary message body exceeds {MAX_BINARY_BODY_BYTES} bytes")
            return data
        descriptor = _binary_descriptor(self.body)
        return None if descriptor is None else _decode_binary_descriptor(descriptor)

    def get_raw(self):
        return self.raw

    def make_request(self, reply_to: str = None) -> str:
        if self.header is None:
            self.header = MessageHeader("None", "None")
            logger.warning("Attempting to make request from message with no header")
        return self.header.make_request(reply_to)

    def set_correlation_id(self, correlation_id: str):
        if self.header is None:
            self.header = MessageHeader("None", "None", correlation_id)
        else:
            self.header.set_correlation_id(correlation_id)

    @staticmethod
    def from_object(msg_contents):
        message = Message()
        logger.debug("In Message.from_object")

        if isinstance(msg_contents, dict):
            logger.debug(f"Message contents: {msg_contents}")
            if "header" in msg_contents:
                message.header = MessageHeader.from_dict(msg_contents["header"])
            if "identity" in msg_contents:
                # Lenient: a malformed identity yields None + a WARN and the message
                # still delivers (UNS-CANONICAL-DESIGN §1.2).
                message.identity = MessageIdentity.from_dict(msg_contents["identity"])
            if "tags" in msg_contents:
                message.tags = MessageTags.from_dict(msg_contents["tags"])
            if "body" in msg_contents:
                message.body = msg_contents["body"]
            if not any(key in msg_contents for key in ["header", "identity", "tags", "body"]):
                logger.debug("Dict contained raw data: Assigning to raw")
                message.raw = msg_contents
        else:
            logger.debug("Message not instance of dict, assigning to raw")
            message.raw = msg_contents

        return message

File: libs/java/src/main/java/com/mbreissi/edgecommons/messaging/Message.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.edgecommons.messaging;

import com.mbreissi.edgecommons.config.ConfigManager;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.JsonSyntaxException;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.Base64;
import java.util.Map;


/**
 * Represents a message in the Greengrass messaging system.
 * Contains message headers and payload data for communication between components
 * and with AWS IoT Core.
 */
public class Message
{
    protected static final Logger LOGGER = LogManager.getLogger(Message.class);
    public static final int MAX_BINARY_BODY_BYTES = 64 * 1024;
    private static final String BINARY_BODY_KEY = "_edgecommonsBinary";
    private static final String BINARY_ENCODING = "base64";

    /** Default serializer: omits null members (Gson default), used for POJO/record/List payloads. */
    private static final Gson DEFAULT_GSON = new Gson();
    /**
     * Null-serializing serializer used only for {@link Map}-shaped payloads, where a present key with
     * a null value is unambiguous intent ({@code map.put("k", null)}) and must serialize as JSON
     * {@code null} — at parity with a Python {@code dict} {@code None}, a TS object {@code null}, and
     * serde. POJO/record payloads deliberately keep the default omit-null behavior (an unset field is
     * ambiguous), so this is not enabled globally (#15).
     */
    private static final Gson NULL_SERIALIZING_GSON = new GsonBuilder().serializeNulls().create();

    MessageHeader header;
    MessageIdentity identity;
    MessageTags tags;
    Object body;
    Object raw;

    /**
     * Private constructor for creating empty messages.
     * Messages should be created using the build methods.
     */
    Message()
    {
        header = null;
        identity = null;
        tags = null;
        body = null;
        raw = null;
    }

    /**
     * Converts the message to a JsonObject representation.
     *
     * @return JsonObject containing the full message data
     */
    public JsonObject toDict()
    {
        JsonObject retVal = new JsonObject();
        if (raw == null)
        {
            if (header != null)
                retVal.add("header", header.toDict());
            // Canonical envelope member order: header, identity, tags, body.
            if (identity != null)
                retVal.add("identity", identity.toDict());
            if (tags != null)
                retVal.add("tags", new Gson().toJsonTree(tags.toDict()).getAsJsonObject());
            retVal.add("body", toJsonElement(body));
        }
        else
        {
            retVal.add("raw", toJsonElement(raw));
        }
        return retVal;
    }

    /**
     * Coerces a message body/raw value to a Gson {@link JsonElement} for serialization. A value that
     * is already a {@link JsonElement} is returned as-is; {@code null} becomes {@link JsonNull}; a
     * {@code byte[]} is encoded as the first-class bounded binary body marker; a {@link Map} is converted with
     * null-valued entries preserved as JSON {@code null} (#15); any other object (POJO, {@code List},
     * primitive wrapper, etc.) is converted via Gson's default reflective adapter (which omits null
     * members). This lets callers pass a plain {@code Map}/POJO to
     * {@link MessageBuilder#withPayload(Object)} and have it serialize correctly, instead of failing
     * with a {@code ClassCastException} at publish time — at parity with the Rust/Python/TypeScript
     * libraries, which accept native maps/objects as payloads.
     *
     * @param value the body or raw value (any type, may be {@code null})
     * @return the value as a {@link JsonElement}
     */
    private static JsonElement toJsonElement(Object value)
    {
        if (value == null)
            return JsonNull.INSTANCE;
        if (value instanceof JsonElement element)
            return element;
        if (value instanceof byte[] bytes)
            return binaryBodyJson(bytes);
        if (value instanceof Map<?, ?>)
            // #15: a Map carries key presence, so a null value is an explicit JSON null (not an
            // ambiguous unset POJO field) — serialize it as such, matching Python/TS/serde. We go via
            // a JSON string (not toJsonTree) because Gson's JsonTreeWriter drops null values in NESTED
            // maps; the string path with serializeNulls preserves them at every nesting level.
            return JsonParser.parseString(NULL_SERIALIZING_GSON.toJson(value));
        return DEFAULT_GSON.toJsonTree(value);
    }

    private static JsonObject binaryBodyJson(byte[] bytes)
    {
        if (bytes.length > MAX_BINARY_BODY_BYTES)
            throw new IllegalArgumentException(
                "Binary message body exceeds " + MAX_BINARY_BODY_BYTES + " bytes");
        JsonObject descriptor = new JsonObject();
        descriptor.addProperty("encoding", BINARY_ENCODING);
        descriptor.addProperty("length", bytes.length);
        descriptor.addProperty("data", Base64.getEncoder().encodeToString(bytes));
        JsonObject marker = new JsonObject();
        marker.add(BINARY_BODY_KEY, descriptor);
        return marker;
    }

    private static JsonObject binaryDescriptor(Object value)
    {
        if (value instanceof JsonObject obj && obj.has(BINARY_BODY_KEY))
        {
            if (!obj.get(BINARY_BODY_KEY).isJsonObject())
                throw new IllegalArgumentException("Binary message body marker must be an object");
            return obj.getAsJsonObject(BINARY_BODY_KEY);
        }
        return null;
    }

    private static byte[] decodeBinaryDescriptor(JsonObject descriptor)
    {
        if (!descriptor.has("encoding") || !BINARY_ENCODING.equals(descriptor.get("encoding").getAsString()))
            throw new IllegalArgumentException("Binary message body encoding must be base64");
        if (!descriptor.has("length") || !descriptor.get("length").isJsonPrimitive())
            throw new IllegalArgumentException("Binary message body length is required");
        int declaredLength = descriptor.get("length").getAsInt();
        if (declaredLength < 0 || declaredLength > MAX_BINARY_BODY_BYTES)
            throw new IllegalArgumentException(
                "Binary message body exceeds " + MAX_BINARY_BODY_BYTES + " bytes");
        if (!descriptor.has("data") || !descriptor.get("data").isJsonPrimitive())
            throw new IllegalArgumentException("Binary message body data is required");
        byte[] decoded = Base64.getDecoder().decode(descriptor.get("data").getAsString());
        if (decoded.length != declaredLength)
            throw new IllegalArgumentException("Binary message body length does not match decoded data");
        return decoded;
    }

    @Override
    public String toString()
    {
        return toDict().toString();
    }

    /**
     * Gets the correlation ID associated with this message.
     *
     * @return The message correlation ID
     */
    public String getCorrelationId()
    {
        if (header == null)
            return null;
        return header.getCorrelationId();
    }

    /**
     * Gets the header information for this message.
     *
     * @return The MessageHeader object
     */
    public MessageHeader getHeader()
    {
        return header;
    }

    /**
     * Gets the UNS identity element of this message ({@code hier}/{@code path}/{@code component}/
     * {@code instance}), or {@code null} when the message carries none (raw messages, messages
     * built without a config-bound builder, or a malformed inbound identity).
     *
     * @return The MessageIdentity object, or {@code null}
     */
    public MessageIdentity getIdentity()
    {
        return identity;
    }

    /**
     * Gets the tags associated with this message.
     *
     * @return The MessageTags object
     */
    public MessageTags getTags()
    {
        return tags;
    }

    /**
     * Adds a tag to this message.
     *
     * @param key The tag key
     * @param value The tag value
     */
    public void injectTag(String key, String value)
    {
        if (tags == null)
            tags = new MessageTags();
        tags.injectTag(key, value);
    }

    /**
     * Gets the message payload body.
     *
     * @return The message payload object
     */
    public Object getBody()
    {
        return body;
    }

    /**
     * Returns true when the payload is a first-class binary body.
     *
     * @return true for an outbound byte[] payload or an inbound binary marker object
     */
    public boolean isBinaryBody()
    {
        return body instanceof byte[] || (body instanceof JsonObject obj && obj.has(BINARY_BODY_KEY));
    }

    /**
     * Decodes the first-class binary message body.
     *
     * @return the decoded bytes, or null when the body is not binary
     * @throws IllegalArgumentException when the inbound binary marker is malformed or too large
     */
    public byte[] getBinaryBody()
    {
        if (body instanceof byte[] bytes)
        {
            if (bytes.length > MAX_BINARY_BODY_BYTES)
                throw new IllegalArgumentException(
                    "Binary message body exceeds " + MAX_BINARY_BODY_BYTES + " bytes");
            return bytes.clone();
        }
        JsonObject descriptor = binaryDescriptor(body);
        return descriptor == null ? null : decodeBinaryDescriptor(descriptor);
    }

    /**
     * Gets the raw message content.
     *
     * @return The raw message object if present, null otherwise
     */
    public Object getRaw()
    {
        return raw;
    }

    public String makeRequest()
    {
        return makeRequest(null);
    }

    /**
     * Prepares this message as a request, setting up correlation and reply information.
     *
     * @param replyTo The topic to send replies to, or null for auto-generated topic
     * @return The correlation ID for tracking the request
     */
    public String makeRequest(String replyTo)
    {
        if (header == null)
        {
            header = new MessageHeader("None", "None", null, null, null, null);
            LOGGER.warn("Attempting to make request from message with no header");
        }
        return header.makeRequest(replyTo);
    }

    /**
     * Sets the correlation ID for this message.
     *
     * @param correlationId The correlation ID to use
     */
    public void setCorrelationId(String correlationId)
    {
        if (header == null)
            header = new MessageHeader("None", "None", correlationId, null, null, null);
        else
            header.setCorrelationId(correlationId);
    }

    /**
     * @deprecated Use {@link MessageBuilder#create(String, String)} instead
     */
    @Deprecated
    public static Message buildFromConfig(String name, String version, Object payload,
                                          ConfigManager configManager)
    {
        return MessageBuilder.create(name, version)
            .withPayload(payload)
            .withConfig(configManager)
            .build();
    }

    /**
     * @deprecated Use {@link MessageBuilder#create(String, String)} instead
     */
    @Deprecated
    public static Message buildFromConfig(String name, String version, Object payload,
                                          ConfigManager configManager, String correlationId)
    {
        Message retVal = new Message();
        MessageHeaderBuilder headerBuilder = MessageHeaderBuilder.create(name, version);
        if (correlationId != null) {
            headerBuilder.withCorrelationId(correlationId);
        }
        retVal.header = headerBuilder.build();
        retVal.tags = MessageTags.fromConfig(configManager);
        if (payload instanceof String payloadStr)
        {
            try
            {
                Gson gson = new Gson();
                // check if a "stringified" json object and convert to object if so
                retVal.body = gson.fromJson(payloadStr, Object.class);
            }
            catch (JsonSyntaxException e)
            {
                // just a regular string
                retVal.body = payloadStr;
            }
        }
        else
        {
            retVal.body = payload;
        }
        return retVal;
    }

    /**
     * Builds a message from a generic message contents object.
     *
     * @param msgContents The content to create the message from
     * @return A new Message instance
     * @deprecated Use {@link MessageBuilder#fromObject(Object)} instead
     */
    @Deprecated
    public static Message build(Object msgContents)
    {
        Message retVal = new Message();
        LOGGER.trace("In Message::build");
        if (msgContents instanceof JsonObject msgJsonObj)
        {
            LOGGER.trace("Message contents: {}", msgJsonObj);
            if (msgJsonObj.has("header"))
            {
                LOGGER.trace("processing header");
                retVal.header = MessageHeader.fromDict(msgJsonObj.getAsJsonObject("header"));
                LOGGER.trace("header deserialized");
            }
            if (msgJsonObj.has("identity"))
            {
                LOGGER.trace("processing identity");
                retVal.identity = parseIdentity(msgJsonObj.get("identity"));
                LOGGER.trace("identity deserialized");
            }
            if (msgJsonObj.has("tags"))
            {
                LOGGER.trace("processing tags");
                retVal.tags = MessageTags.fromDict(msgJsonObj.getAsJsonObject("tags"));
                LOGGER.trace("source deserialized");
            }
            if (msgJsonObj.has("body"))
            {
                LOGGER.trace("processing body");
                retVal.body = msgJsonObj.get("body");
                LOGGER.trace("body desiralized");
            }
            if (!(msgJsonObj.has("header") || msgJsonObj.has("identity")
                    || msgJsonObj.has("tags") || msgJsonObj.has("body")))
            {
                LOGGER.trace("Json contained raw string: Assigning to raw");
                retVal.raw = msgJsonObj;
            }
        }
        else
        {
            LOGGER.trace("Message not instance of JsonObject, assigning to raw");
            retVal.raw = msgContents;
        }
        return retVal;
    }

    /**
     * Leniently parses an envelope {@code identity} member: a non-object element or a malformed
     * identity yields {@code null} plus a WARN (the message still delivers). Shared by the
     * deprecated {@link #build(Object)} and {@link MessageBuilder#fromObject(Object)}.
     *
     * @param identityElement the raw {@code identity} JSON element
     * @return the parsed identity, or {@code null} when malformed
     */
    static MessageIdentity parseIdentity(JsonElement identityElement)
    {
        if (identityElement == null || !identityElement.isJsonObject())
        {
            LOGGER.warn("Malformed message identity: 'identity' is not an object; dropping identity");
            return null;
        }
        return MessageIdentity.fromDict(identityElement.getAsJsonObject());
    }

}
